Exclude only log4j2 queue from class loader

Minecrell opened the pull request:
Right now the complete `com.mojang.*` package is excluded from the `LaunchClassLoader` on the server. This means mods can't transform `authlib` for example which doesn't need to be excluded for the UI to work. By excluding only the specific log4j2 `QueueLogAppender`, mods can also transform the classes in the `com.mojang.authlib` package.

Specifically we want to use this in the Sponge mod to implement our [`GameProfile` interface](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameProfile.java) into Mojang's `GameProfile` class (`com.mojang.authlib.GameProfile`). This isn't possible right now because the class is excluded from the `LaunchClassLoader` and can't be transformed because of that.

I have tested that the UI is still printing the log correctly using this change.

LexManos commented:
Although the intent is for log4j, i'm rather curious if I want to allow the editing of authlib, as that seems like a rather bad idea.
Could your system not be done using a simple wrapper class in this instance?

Minecrell commented:
I had started implementing it using wrappers, but it made things in some places quite complicated. We provide a mutable `List<sponge.GameProfile>` for editing the player list sent together with the status ping response. This is a `authlib.GameProfile[]` normally, but we've changed it to a `List<authlib.GameProfile>` so new items can be removed and added.

Now if we can patch `authlib.GameProfile` with our interface directly, all we need to do is to cast our list, because we don't need to convert all the profiles to our wrapper: https://github.com/Minecrell/Sponge/commit/01f388323c6f267c0d6c723ed475176c70ce8725#diff-fc23d1490b23b8e7371ff44ca56fb779R69

To replace this using wrappers you would need a custom `List` implementation that would convert from the wrapper to the `authlib.GameProfile` on every operation, for example like this: https://gist.github.com/Minecrell/bed8e8f3abf2201f2afa
Compared to the other version, this requires more effort to maintain and is also pretty inefficient.

On the other hand, patching `com.mojang.authlib` classes is **already possible on the client**. Only the server is preventing it through this class loader exclusion rule. So if you would want to disallow patching these, you would need to disable it in the client first.

LexManos commented:
I still don't like opening up modification of authlib, but whatever.
As a future note: 'But it works on the client!!!!' is not a valid excuse, it would just mean I have to lock it down on the client -.-

Minecrell commented:
I said you would need to prevent it on the client first, if you would want to prevent modification of authlib. It was not meant as excuse.

Thanks for merging.

Minecrell commented:
@LexManos Do you need to update the submodule in MinecraftForge/MinecraftForge so we can use the updated FML version in Sponge? If so, would you mind doing this in the next days? Thanks!

Mumfrey commented:
@Minecrell 

> To replace this using wrappers you would need a custom List implementation that would convert from the wrapper to the authlib.GameProfile on every operation

I realise this is moot now, but just for future reference you can use guava's `Lists.transform` for this. It provides essentially a transform-on-access *view* of an existing list via a `Function` object passed to the method. For example say you have a `List<Foo>` and you'd like a `List<String>` which is a *view* of the original list, you can do

```java
List<Foo> listOfFoos = ...;
List<String> listOfStrings = Lists.transform(listOfFoos, Functions.toStringFunction());
```

which uses the built-in function `toStringFunction` but you can easily write your own.

Anyway, just something useful to bear in mind.

Minecrell commented:
@Mumfrey Yeah, I would have used it but it's only a view so it doesn't support adding or removing elements. Although Guava has the `Converter` class additionally which provides convertion in both ways they do not have a list implementation which can convert _in both ways_, and would also support adding new elements through that.

Mumfrey commented:
Well you can add/remove elements to the underlying list, you just can't add them to the view. But yes, I was really just pointing out that there are off-the-shelf classes for exactly what is needed, so if the mixing of gameprofile ever *does* become an issue then there are still options open which (though sub-optimal) are still somewhat workable.
